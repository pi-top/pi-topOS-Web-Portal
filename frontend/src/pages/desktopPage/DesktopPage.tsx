import React from "react";
import cx from "classnames";

import styles from "./DesktopPage.module.css";
import Spinner from "../../components/atoms/spinner/Spinner";


export type Props = {
  url: string
  error: boolean
}

export const errorMessage = "There was an error opening the page. Make sure VNC is enabled in your device and refresh this page.";

export default ({ url, error} : Props) => {
  return (
    <div className={cx(styles.content)}>
      <div className={styles.container}>
        { url === "" && error && <span className={styles.error}>{errorMessage}</span> }
        { url === "" && !error && <Spinner size={80} /> }
        { url !== "" &&
          <iframe
            src={url}
            data-testid="vnc-desktop"
            title="Desktop"
            className={styles.frame}
          >
          </iframe>
        }
      </div>
    </div>
  );
};
