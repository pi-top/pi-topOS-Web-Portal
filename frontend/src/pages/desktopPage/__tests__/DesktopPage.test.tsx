import React from "react";
import {
  render,
  fireEvent,
  BoundFunction,
  GetByBoundAttribute,
  QueryByText,
  GetByText,
  wait,
  screen,
} from "@testing-library/react";

import DesktopPage, { Props, errorMessage } from "../DesktopPage";
import querySpinner from "../../../../test/helpers/querySpinner";


describe("DesktopPage", () => {
  let desktopPage: HTMLElement;
  let defaultProps: Props;
  let queryByAltText: any;
  let queryByText: any;
  let queryByLabelText: any;
  let queryByTestId: any;
  let getByText: any;
  let rerender: any;

  beforeEach(() => {
    defaultProps = {
      url: "",
      error: false
    };

    ({
      container: desktopPage,
      queryByAltText,
      queryByText,
      queryByLabelText,
      queryByTestId,
      getByText,
      rerender,
    } = render(<DesktopPage {...defaultProps} />));
  })

  it("renders a spinner on mount", () => {
    expect(querySpinner(desktopPage)).toBeInTheDocument();
  });

  describe("on error", () => {
    beforeEach(() => {
      defaultProps = {
        url: "",
        error: true,
      };

      rerender(<DesktopPage {...defaultProps} />);
    })

    it("renders the correct message", async () => {
      expect(await screen.findByText(errorMessage)).toBeInTheDocument();
    });

    it("doesn't render a spinner", () => {
      expect(querySpinner(desktopPage)).not.toBeInTheDocument();
    });

  });

  describe("when an URL is received", () => {
    beforeEach(() => {
      defaultProps = {
        url: "http://pi-top.com",
        error: true,
      };

      rerender(<DesktopPage {...defaultProps} />);
    })

    it("renders an iframe", () => {
      expect(queryByTestId("vnc-desktop")).toBeInTheDocument();
    });

    it("iframe opens the provided url", () => {
      expect(queryByTestId("vnc-desktop")).toMatchSnapshot();
    });

    it("doesn't render a spinner", () => {
      expect(querySpinner(desktopPage)).not.toBeInTheDocument();
    });

    it("doesn't render an error message", async () => {
      expect(screen.queryByText(errorMessage)).not.toBeInTheDocument();
    });

  });

});
