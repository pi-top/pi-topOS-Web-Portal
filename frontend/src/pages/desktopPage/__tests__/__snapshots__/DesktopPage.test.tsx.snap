// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DesktopPage when an URL is received iframe opens the provided url 1`] = `
<iframe
  class="frame"
  data-testid="vnc-desktop"
  src="http://pi-top.com"
  title="Desktop"
/>
`;
